Restructure repo by introducing account creation and authorization in… (#6)
Sitelink3D-v2-Developer/sitelink3dv2-api-documentation: security/security_overview.md

```
api.giantbank.com/generate_statement "bob@tinybank.com" "password"
```

> _Example Representation of a Band Customer Using Identity_
> 
> | BOB                 |
> |---------------------|
> | **Bank**            |
> | Tiny Bank           |
> | **Identity**        |
> | bob@tinybank.com    |
> | **Capabilities**    |
> | Generate statements |

This process is required because the service that generates statements needs to first determine whether the requester is authorized to do so. Giant Bank's software hence determines whether the user account exists by looking up bob@tinybank.com in a database that Giant Bank maintains, it checks that the provided password is correct, runs the statement generation if so and then returns a result that is displayed on Tiny Bank's website. 

The same process occurs for all services that have been on-sold and for all customers of not only Tiny Bank, but all the other new institutions that have signed up to use Giant Bank's platform.

> _Some Useful Definitions_
> 
> **Authentication**
>
> The process of determining who you are.
> 
> **Authorization**
>
> The process of determining that you have the right to access a resource.

> One can perhaps start to imagine some of the problems that rapidly arise from this approach. Because Giant Bank is required to authenticate the identities of users of its services, it necessarily needs to know about all the customers of all the banks that is on-sells to. Privacy issues aside, the mechanics of this become unmanageable.

Reviewer: ```suggestion
One can perhaps start to imagine some of the problems that rapidly arise from this approach. Because Giant Bank is required to authenticate the identities of users of its services, it necessarily needs to know about all the customers of all the banks that it on-sells to. Privacy issues aside, the mechanics of this become unmanageable.
```

> 3. The Client sends the set of signed Permissions to the Server together with a request.
> 4. The Server uses the set of permissions as before.

In the previous section we listed five shortcomings of IBAC. Here we present the five solutions provided by ZBAC.

> _Benefits of ZBAC_
> 
> 1. **Identity information is contained** within domain boundaries as identity is authenticated locally in order to issue anonymous authorization tokens.
> 2. **It is easy to manage** fine-grained authorization models because policy information is local to the business or domain closest to the user.
> 3. **Consistent resource naming** is achieved because the token is an abstraction that allows separation of inter-business and intra-business concerns.
> 4. **Permission delegation is easy to achieve** by controlling policy and selectively distributing anonymous authorization tokens.
> 5. **There is no vulnerability** to confused deputy attacks as decisions are always based on the possession of valid tokens rather than identities requiring authentication.

## ZBAC In Action Within Sitelink3D v2

> Already we have presented enough information to make some sense of the Sitelink3D v2 REST API defined on our [Swagger](https://api.sitelink.topcon.com/swagger/) page. Security tokens are provided to Sitelink3D v2 via the X-Topcon-Auth header parameter.

Reviewer: ```suggestion
Already we have presented enough information to make some sense of the Sitelink3D v2 REST API defined on our [Swagger](https://api.sitelink.topcon.com/swagger/) page. Security tokens are provided to Sitelink3D v2 via the Authorization header parameter.
```

Reviewer: We should update swagger to drop X-Topcon-Auth

This limitation can be addressed with a modification to how operations are defined within Nodes and within JWTs. What we have previously referred to as an action list needs to become a list of service to action mappings. This design allows individual services to be specified and for each service, the specific actions that can be performed on the resource. This is the purpose of the _"Actions"_ field which is abbreviated to _"act"_. Naturally, the "*" wildcard can be used to introduce flexibility in how actions are expressed.

Returning to our banking example, it makes sense that one of the banking services is responsible for managing clients in the bank database. Let's call this service _"client_manager"_. The client_manager service should be able to call the _"add_client"_ and _"rename_client"_ functions on the bank's client resources. It may be the case that all services are able to list the bank's clients. Such an action set would be expressed using our new syntax as follows.

```
"act": {
  "client_manager":["add_client", "rename_client"],
  "*":["list_clients"]
}
```

> Let's add a modified subject and collectively define a exactly what the client_manager service can do to all clients within Loans Are Us. This provides a complete picture of what this permission is specifying. A possible JWT may look like the following.

Reviewer: Might want to re-frame this slightly to be actions the bearer of the token can expect a service to perform on their behalf. The service can do what ever it likes to the resources under its control

Reviewer 2: Agreed.

```
{
  "iss":"ad8d2c4049b243cabffa14968b5a54fa",
  "sub":"urn:X-topcon:le:564529a7-3774-4e12-a414-27efb60b8214:bank:9b178e64-322c-4f23-9252-bd3b6c96823c:members:clients",
  "scope": "*",
    "act": {
      "client_manager":["add_client", "rename_client"],
      "*":["list_clients"]

>     }

Reviewer: different tab sizes

> _Improved Definitions for Sitelink3D v2 Security Node Properties_
> 
> **Permission**
> 
> Permissions are defined by a collection of common fields present in both a Node and in the payload of a signed JWT. The fields that collectively comprise a permission are the subject "sub", a scope that identifies whether the subject is restricted in some way "scp" and a set of service, action pairings "act".
> 
> A JWT contains only a single permission. Multiple permissions are achieved by issuing multiple JWTs. A Node will by default contain a permission that allows it to issue unrestricted JWTs for its unique subject. Detailed information on how permission fields work within Sitelink3D v2 is available in the Swagger documentation.
> 
> **Policy**
> 
> A policy is a set of permissions. As JWTs can only contain a single permission, policies are exclusive to Nodes. A Node is likely to be responsible for multiple resources, or even different scopes of the same resource chain. This is achieved by simply storing a collection of one or more permissions in a list called a policy. > A node must have at least one permission for it to generate JWTs against.

> Armed with these new definitions, we are able to express the Loans Are Us Node policy described above as the following policy comprised of two permissions. The first permission states that the Node is free to chain any number of resource identifier strings to the subject that represents its own resource space and generate JWTs for any services and actions to operate on that space. The second it the permission delegated by Giant Bank that allows for JWTs to be created that control access to the subset of Giant Bank's resource space as defined by the B2B relationship between the banks.

Reviewer: ```suggestion
Armed with these new definitions, we are able to express the Loans Are Us Node policy described above as the following policy comprised of two permissions. The first permission states that the Node is free to chain any number of resource identifier strings to the subject that represents its own resource space and generate JWTs for any services and actions to operate on that space. The second is the permission delegated by Giant Bank that allows for JWTs to be created that control access to the subset of Giant Bank's resource space as defined by the B2B relationship between the banks.
```

When designing the creation of Nodes and Node policies then, there is a balance that can be struck. When looking to restrict access, at one end of the spectrum a Node policy can be very broad and the JWTs issued by that Node then need to contain the appropriate restrictions in terms of explicit narrowing of the subject, scope and action fields or some combination thereof.

Alternatively, a Node may be created with quite a specific policy that grants it only the ability to issue JWTs for specific services, subjects and scopes. In this case, the Node may then issue unrestricted JWTs that have the effect of resolving to the policies of the Node when intersected by Trust Manager. Of course, any combination of these extremes is also possible. Such scenarios are visualized below. The contents of the policy and JWT can be reversed and still result in the same effective permissions.

**Policy**

```
[
  {
    "sub": "urn:X-topcon:le:564529a7-3774-4e12-a414-27efb60b8214",
    "scp": "*",
    "act": { "*":["*"] }
  }
]
```

**JWT**

```
{
  "sub": "urn:X-topcon:le:564529a7-3774-4e12-a414-27efb60b8214:members:clients:account:12345678",
  "scp": "node",

>   "act": &nbsp;{

Reviewer: ```suggestion
  "act": {
```

    "account_service":["view_balance", "deposit", "transfer"]
  }
}
```

**Effective Permissions**

> ```

Reviewer: ```suggestion
```json
```

Reviewer: we should do this for all code sections

Reviewer 2: Done.

{
  "name":"Loans Are Us",
  "sub": "urn:X-topcon:le:564529a7-3774-4e12-a414-27efb60b8214:members:clients:account:12345678",
  "scope": "node",

>   "act": &nbsp;{

Reviewer: ```suggestion
  "act": {
```
